### PR TITLE
Fix: Angie support for required children [ED-22527]

### PR DESCRIPTION
**Pull request by asafdl**
## Live Playground Preview
<a href="https://playground.wordpress.net?blueprint-url=data:application/json,%7B%0A%20%20%20%20%22%24schema%22%3A%20%22https%3A%2F%2Fplayground.wordpress.net%2Fblueprint-schema.json%22%2C%0A%20%20%20%20%22landingPage%22%3A%20%22%2Fwp-admin%22%2C%0A%20%20%20%20%22preferredVersions%22%3A%20%7B%20%22php%22%3A%20%227.4%22%2C%20%22wp%22%3A%20%22latest%22%20%7D%2C%0A%20%20%20%20%22features%22%3A%20%7B%0A%20%20%20%20%20%20%20%20%22networking%22%3A%20true%0A%20%20%20%20%7D%2C%0A%20%20%20%20%22extraLibraries%22%3A%20%5B%20%22wp-cli%22%20%5D%2C%0A%20%20%20%20%22steps%22%3A%20%5B%0A%20%20%20%20%20%20%20%20%7B%20%22step%22%3A%20%22login%22%2C%20%22username%22%3A%20%22admin%22%2C%20%22password%22%3A%20%22password%22%20%7D%2C%0A%20%20%20%20%20%20%20%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%22step%22%3A%20%22defineWpConfigConsts%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%22consts%22%3A%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22ELEMENTOR_SHOW_HIDDEN_EXPERIMENTS%22%3A%20true%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%2C%0A%20%20%20%20%20%20%20%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%22step%22%3A%20%22installPlugin%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%22pluginData%22%3A%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22resource%22%3A%20%22url%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22url%22%3A%20%22https%3A%2F%2Fgithub.com%2Felementor%2Felementor%2Freleases%2Fdownload%2Fci-artifacts%2Fpr-35903-821ee6aac6cf8db02950f3ed7d3cc97d0e57407a.zip%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%2C%0A%20%20%20%20%20%20%20%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%22step%22%3A%20%22installTheme%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%22themeData%22%3A%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22resource%22%3A%20%22wordpress.org%2Fthemes%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22slug%22%3A%20%22hello-elementor%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%2C%0A%20%20%20%20%20%20%20%20%7B%20%22step%22%3A%20%22activateTheme%22%2C%20%22themeFolderName%22%3A%20%22hello-elementor%22%20%7D%2C%0A%20%20%20%20%20%20%20%20%7B%20%22step%22%3A%20%22wp-cli%22%2C%20%22command%22%3A%20%22wp%20option%20add%20&#039;elementor_onboarded&#039;%201%22%20%7D%2C%0A%20%20%20%20%20%20%20%20%7B%20%22step%22%3A%20%22wp-cli%22%2C%20%22command%22%3A%20%22wp%20elementor%20experiments%20activate%20e_atomic_elements%2Ccontainer%2Ce_opt_in_v4%2Ce_classes%22%20%7D%2C%0A%20%20%20%20%20%20%20%20%7B%20%22step%22%3A%20%22updateUserMeta%22%2C%20%22meta%22%3A%20%7B%20%22announcements_user_counter%22%3A%20999%2C%20%22elementor_enable_ai%22%3A%200%2C%20%22_e_welcome_popover_displayed%22%3A%201%20%7D%2C%20%22userId%22%3A%201%20%7D%0A%20%20%20%20%5D%0A%7D" target="_blank" rel="noopener noreferrer">
  Open WordPress Playground Preview
</a>

## 1. Problem & Context

Form elements (success/error messages) need required child enforcement in Angie to ensure valid XML composition. Currently missing default paragraph children and validation to prevent incomplete form structures.

## 2. What Changed (Where)

| Layer | Changes |
|-------|---------|
| **PHP (Atomic Widgets)** | Added `meta['required']` flag to form message builders; implemented `define_default_children()` & `get_default_status_paragraph_text()` abstract methods with localized defaults |
| **Builders** | Extended `Element_Builder` and `Widget_Builder` with `meta()` method & meta serialization in `build()` |
| **TypeScript/Composition** | New `RequiredChildrenEnforcer` validates XML has all required children; `getRequiredDefaultChildTypes()` extracts required tag names from config |
| **Schema/LLM** | Added `required_direct_children` to widget schema guidance; updated V1 model types to include `meta` field |
| **Tests** | Added 5 test suites covering enforcer logic, schema extraction, and full composition build scenarios |

## 3. How It Works

When building composition from XML, `CompositionBuilder` now filters widgets with `default_children` marked `meta.required`, then runs `RequiredChildrenEnforcer` to validate all required tags exist as direct children. Enforcer traverses XML, collects missing tags per element type, throws descriptive error with configuration-id if any missing. Form widgets generate default `Atomic_Paragraph` children with localized text at instantiation; meta flag propagates through element builder → final JSON model for validation.

## 4. Risks

**Import ordering**: PHP uses new imports for control types & HTML props—verify no circular dependencies with Atomic_Paragraph. **Schema coupling**: LLM guidance now depends on `default_children` config; schema must stay in sync with PHP defaults or LLM generates invalid XML. **Enforcer edge cases**: Only validates direct children by tag name—if widget type appears in multiple places (nested), false positives possible (mitigate via test coverage, which is present).

_Generated by LinearB AI and added by gitStream._
<sub>AI-generated content may contain inaccuracies. Please verify before using.
💡 **Tip:** You can customize your AI Description using **Guidelines** [Learn how](https://docs.gitstream.cm/automation-actions/#describe-changes)</sub>

### CODE REVIEW

File: packages/packages/core/editor-canvas/src/composition-builder/__tests__/composition-builder.test.ts [hunk 1]
```diff
@@ -188,3 +188,189 @@ describe( 'CompositionBuilder.build applyProperties after create', () => {
...
+				elType: 'e-form-error-message',
+			},
+			'e-form-input': { title: 'Input', controls: {}, elType: 'widget' },
+			'e-form-success-message': {
+				title: 'Success',
+				controls: {},
+				elType: 'e-form-success-message',
+			},
+		} as Record< string, V1ElementConfig >;
+		const builder = CompositionBuilder.fromXMLString(
```
Comment (ronenboxer) [inline packages/packages/core/editor-canvas/src/composition-builder/__tests__/composition-builder.test.ts]: These all feel like a const

File: packages/packages/core/editor-canvas/src/composition-builder/__tests__/composition-builder.test.ts [hunk 2]
```diff
@@ -188,3 +188,189 @@ describe( 'CompositionBuilder.build applyProperties after create', () => {
...
+		// Act & Assert
+		await expect( builder.build( createMockRootContainer() ) ).rejects.toThrow(
+			/Missing required direct child element tag\(s\): e-form-success-message, e-form-error-message/
+		);
+		expect( createElementMock ).not.toHaveBeenCalled();
+	} );
+
+	it( 'rejects build when only some required direct children exist', async () => {
+		// Arrange
+		let elementIdSequence = 0;
```
Comment (ronenboxer) [inline packages/packages/core/editor-canvas/src/composition-builder/__tests__/composition-builder.test.ts:255]: and a global var to be initialized in the before each hook

File: packages/packages/core/editor-canvas/src/composition-builder/utils/required-default-child-tags.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import { type V1ElementConfig } from '@elementor/editor-elements';
+
+export type DefaultChildTemplate = Record< string, unknown >;
+
+export function resolveDefaultChildTemplateTagName( template: DefaultChildTemplate ): string {
```
Comment (ronenboxer) [inline packages/packages/core/editor-canvas/src/composition-builder/utils/required-default-child-tags.ts]: tag names or instance template type?

File: packages/packages/core/editor-canvas/src/composition-builder/utils/required-default-child-tags.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import { type V1ElementConfig } from '@elementor/editor-elements';
+
+export type DefaultChildTemplate = Record< string, unknown >;
```
Comment (ronenboxer) [inline packages/packages/core/editor-canvas/src/composition-builder/utils/required-default-child-tags.ts]: Don't we (shouldn't we) have a properly defined type for ChildTemplates?

**Review by ronenboxer — DISMISSED** [review]

**Review by davseve — APPROVED** [review]
